### ISSUE TITLE
HandlerDisconnect crush 

### ISSUE THREAD

**Issue by NoobsEnslaver** [go-gst/go-gst#94]
Hi. I investigated memory leaks in my application, decided to apply recommendations you suggested here, in the issues - I added the removal of the event handler using HandlerDisconnect and started getting critical messages like this
```
(test1:154528): GLib-GObject-CRITICAL **: 15:31:56.550: g_closure_ref: assertion 'closure->ref_count > 0' failed

(test1:154528): GLib-GObject-CRITICAL **: 15:31:56.550: g_closure_unref: assertion 'closure->ref_count > 0' failed
```
and then crush, although I can't figure out what I'm doing wrong - everything looks like in the examples. Here is my code

```go
package main

import (
	"errors"
	"fmt"
	"time"

	"github.com/go-gst/go-glib/glib"
	"github.com/go-gst/go-gst/gst"
)

func main() {
	gstInit()
	go bench()
	go bench()
	go bench()
	go bench()
	go bench()
	go bench()
	go bench()
	go bench()
	go bench()
	go bench()
	go bench()
	go bench()
	go bench()
	go bench()
	bench()
}

func bench() {
	for {
		if err := test(); err != nil {
			fmt.Println("ERROR: ", err)
			return
		}
		fmt.Print(".")
		// runtime.GC()
	}
}

func gstInit() {
	gst.Init(&[]string{})
	mainLoop := glib.NewMainLoop(glib.MainContextDefault(), false)
	go mainLoop.Run()
}

func test() error {
	var pipeline *gst.Pipeline
	var err error
	strPipeline := `rtpsession name=r
   audiotestsrc
   ! opusenc
   ! rtpopuspay pt=96
   ! application/x-rtp,media=audio,clock-rate=48000,payload=96,encoding-name=OPUS
   ! r.send_rtp_sink
 r.send_rtp_src ! identity name=rtp-in-inspector
   ! udpsink name=udp_rtp_sink host=127.0.0.1 port=37244 async=false
 r.send_rtcp_src
   ! udpsink name=udp_rtcp_sink host=127.0.0.1 port=50388 async=false
`
	if pipeline, err = gst.NewPipelineFromString(strPipeline); err != nil {
		return err
	}
	if !pipeline.GetPipelineBus().AddWatch(handleEvent) {
		return errors.New("can't add watch")
	}
	defer pipeline.GetPipelineBus().RemoveWatch()

	err = pipeline.BlockSetState(gst.StatePlaying)
	if err != nil {
		return err
	}

	inspRTP, err := pipeline.GetElementByName("rtp-in-inspector")
	if err != nil {
		return err
	}

	h, err := inspRTP.Connect("handoff", func(self *gst.Element, buff *gst.Buffer) {})
	if err != nil {
		return err
	}
	defer inspRTP.HandlerDisconnect(h)

	time.Sleep(500 * time.Millisecond)

	return pipeline.BlockSetState(gst.StateNull)
}

func handleEvent(msg *gst.Message) bool {
	// fmt.Println(msg)
	return true
}

```

go.mod:
```
module test1

go 1.22

toolchain go1.22.2

require (
	github.com/go-gst/go-glib v1.0.1
	github.com/go-gst/go-gst v1.0.1-0.20240412193824-b9552df57e79
)

require github.com/mattn/go-pointer v0.0.1 // indirect
```

just compile it, run and wait a bit.

Also - memory still leaking, even if i remove `inspRTP.Connect` and manipulations with Bus at all.

**Comment by RSWilli** [comment]
If I find some time I will look into this. Although a heads up: memory consumption of go programs is a bit magic from my experience. See https://tip.golang.org/doc/gc-guide

Essentially, when your system has enough free RAM, the GC will step back and not collect garbage to improve the performance of your program. On top of that Linux RAM consumption is even more magic. If you want to diagnose the RAM usage of a golang application, always look at the reserved memory (RES in htop).

The warnings you mention above could be caused by the `defer inspRTP.HandlerDisconnect(h)`. In my applications I had some issues with disconnecting handlers when the handler references elements in the pipeline, thus keeping them alive by creating a reference cycle. Your case may differ

**Comment by NoobsEnslaver** [comment]
Problem with `inspRTP.HandlerDisconnect(h)` for sure. With or without `defer` - doesn't matter. There is not just warning - it is crashes after some time with sigfault. 
About memory leaks - it is not for this issue, i'll send some fixes or recommendations in Issues if found something.

**Comment by RSWilli** [comment]
I can confirm that this crashes. It seems to be a race condition related to the `HandlerDisconnect` but all code in there is protected by mutexes: https://github.com/go-gst/go-glib/blob/main/glib/gobject.go#L317-L330

I think the problem lies instead at the connect part of the code, because the creation of the closure is not protected by a mutex: https://github.com/go-gst/go-glib/blob/main/glib/connect.go#L34

I could not find anything yet about the internals of `g_signal_connect_closure`. It might need a mutex protecting it.

My reasoning for it is as follows: If two closures get created with the same ID, then un-reffing will unref the first one twice and leak the second one, as the handle is a key in the map of closures in the bindings.

**Comment by nickeygit** [comment]
Hi guys,

I'm trying to fix this issue - and got a working fix/workaround. The test code in the initial comment - does work (including uncommented runtime.GC() - which makes crashes earlier).
But I need the fix/workaround to be accepted into your repo @RSWilli.
(If you agree - it is fix. If not - it's just my workaround :)

The full fix - is quite big - so I'd like to hear your opinion first before creating nice-clean PR.

I've spent many hours for debugging and the test code has at least two issues:

1. Thread-safety.
I found that the upstream gstreamer library is NOT full thread-safe being called from CGO/multithreaded env (despite gstreamer declares _some_ public functions to be "MT-safe").
I don't know if native C multithreaded app has same issues. Just checked CGO go-gst wrapper.

So, the fix is obvious:
to add a global go-gst lock and make all the public methods (gst.NewPipelineFromString, pipeline.GetPipelineBus, bus.AddWatch, Unref() etc) to hold it during run.
And this does work (added for those mentioned methods used in the test for now).
Skipping locking _any_ method - crashes the app sooner or later. Only locking for all of the used methods - helps.
So, somethind like this must be added into every pub method in go-gst:
```
GlobalLock.Lock()
defer GlobalLock.Unlock()
```
GlobalLock - is declared somewhere in go-gst too.

The problem is: there are no unit-tests in the repo - so I can't say the changes will not affect existing users.
But existing users - are single-threaded (mostly; as I see). So, adding a boolean switch to Lock/Unlock methods of GlobalLock object - can enable or disable the protection. And by default it will be disabled - so existing single-threaded apps will see no functional changes at all.
Multi-threaded apps - will just have to call something like gst.EnableMT() (which will call GlobalLock.Enable() to set the switch on) - and the locking will be enabled.
This, sure, adds some performance decreasing (to near 1-threaded) - but not so much apps does nothing but call GST API. So, for real-world MT apps this should not be noticable.

Also, to make Unref()s being locked too - it's needed to add a forwarding Unref() method to every object (pipeline, bus etc) and change runtime.SetFinalizer(...) to use it, not Object directly. And sure call global locking. Like this:
```
func (e *Element) Unref() {
       GlobalLock.Lock()
       defer GlobalLock.Unlock()

       e.Object.Unref()
}
```

2. Defers (optional).
Using defers with indirect objects - lets GC to free objects before the deferred function runs (!).
Even when last line uses pipeline:
```
return pipeline.BlockSetState(gst.StateNull)
```
as GC can run after return but before running defereds.

The problem is in these train-like calls series:
```
defer pipeline.GetPipelineBus().RemoveWatch()
```
Here, pipeline.GetPipelineBus() is called and returns a bus object. Then bus.RemoveWatch() method value is scheduled to run after the function ends.
Note: if pipeline object is not used in this function anymore - it gets freed by GC (with Unref() calling) - BEFORE deferred funcs/methods run.
Needless to say that gstreamer lib will not be happy seeing calls: pipeline.Unref() and then using the pipeline (via its bus or so). And gstreamer/glib asserts raise.
To fix this - either:
* (not 100% guaranteed) the app have to keep a reference to pipeline (like runtime.KeepAlive(pipeline) at the function bottom) - but it's ugly and don't fully help.
* or app have to defer a wrapper func like:
```
defer func() {
    pipeline.GetPipelineBus().RemoveWatch()
}()
```
So pipeline is referred till this defered is executed and GC will not free it too early. Less ugly, but better add "bus == nil" checking.
* or add a forwarding method RemoveWatch() to pipeline itself - so client apps can call just like:
defer pipeline.RemoveBusWatch()
This - for sure - adds too much knowledge to pipeline object itself, as go-gst is only wrapper for gstreamer and it just should reflect original object tree.
From other side - the app has to have this knowledge, and so - be more complex. If we add this knowledge to go-gst wrapper - we eliminate this complexity for the app. Plus, we can and should add "bus == nil" check into it.
And this way, can be added some more forwarding methods - to simplify wide-used freeing/removing (excluding Unrefs - they should not be called manually).

So, these are proposals for the two issues.
What would you say?

Is global go-gst locking an acceptable method? (at least for the first approximation)
If yes - I'll prepare a PR.

Thanks

**Comment by RSWilli** [comment]
Hey @nickeygit , thanks for suggesting a fix for this. Help is always appreciated :)

I do think that a go-gst wide global lock is not the solution for this issue. We do not need to make every call MT safe, and adding a global lock to the wrong places could lead to some deadlocks (e.g. gstreamer Handler callbacks are called blocking from the streaming thread).

The only part of go-gst that is global, and shared across all pipelines is the signal handler registry. Golang does not allow CGO to call go function pointers, so we need a way to lookup dynamic functions. (This causes a lot of headaches with memory leaks, so if you have a better solution I am open for suggestions)

My gut tells me that the above code needs a mutex to prevent the race condition, but only in the line https://github.com/go-gst/go-glib/blob/main/glib/connect.go#L34 . This however depends on the underlying C implementation. I will ask the gstreamer devs for clarification there.

**Comment by nickeygit** [comment]
Ok.

And what about 2-nd part - defers?
It's another issue - too early objects freeing by GC.

**Comment by RSWilli** [comment]
If you ran into problems there, please create another issue to track it. Although I'd expect that a cleanup of the bus also removes all watches attached to it. Also the following is possible, which delays some finalizers:

```go
pipeline, _ := gst.NewPipelineFromString(strPipeline)
bus := pipeline.GetPipelineBus()
bus.AddWatch(...)

defer bus.RemoveWatch()
```

**Comment by RSWilli** [comment]
> I will ask the gstreamer devs for clarification there.

It may be that the signal handler Ids are only unique per object, which could mean that the current implementation is broken in more ways than expected...

**Comment by sdroege** [comment]
If you can, store everything you need inside the `GClosure` (you can store anything you want in there!) and similarly for e.g. the bus watch store everything you need inside the `user_data`. Using global storage for these things seems overly complicated.

**Comment by RSWilli** [comment]
> you can store anything you want in there!

From a C standpoint this is true. When mixing in go objects then there are some rules, because the "anything" is passed to C.

See https://pkg.go.dev/cmd/cgo#hdr-Passing_pointers

P.S: these rules refer to `runtime.Pinner` (which seems to not fit this use case) and [runtime/cgo.Handle](https://pkg.go.dev/runtime/cgo@go1.22.5#Handle) which essentially stores the object in a global store and returns a key to it.

**Comment by nickeygit** [comment]
> It may be that the signal handler Ids are only unique per object, which could mean that the current implementation is broken in more ways than expected...

You are right.
I've checked Glib - and it uses both object and handle_id to index/search handlers.

gobject/gsignal.c: handler_lookup():
```
  if (handler_id)
    {
      Handler key;
      key.sequential_number = handler_id;
      key.instance = instance;
      return g_hash_table_lookup (g_handlers, &key);
    }
```
So we should have separate signals maps per object too.

**Comment by nickeygit** [comment]
Also, have created a native C equivalent of this test code, multithreaded.
And it is runing just fine, not crashing.

But still slightly mem-leaking.

**Comment by RSWilli** [comment]
> So we should have separate signals maps per object too.

We could also roll our own handler IDs and store them in a map. maybe with https://pkg.go.dev/runtime/cgo@go1.22.5#Handle , I wanted to try that out anyways

**Comment by nickeygit** [comment]
> We could also roll our own handler IDs and store them in a map. maybe with https://pkg.go.dev/runtime/cgo@go1.22.5#Handle , I wanted to try that out anyways

Just wondering if you are going to implement this anytime soon?

Thanks

**Comment by RSWilli** [comment]
I experimented with a change:

I created a global unique `signalIDx` variable that counts up on every attached signal, making our signal handle globally unique.

But then I still got some crashes, I think the way the handlers and closures are unreffed is wrong. The closure is sometimes unreffed more than once:

```

(__debug_bin3648449645:50306): GLib-GObject-CRITICAL **: 18:55:36.501: g_closure_ref: assertion 'closure->ref_count > 0' failed

(__debug_bin3648449645:50306): GLib-GObject-CRITICAL **: 18:55:36.502: g_closure_unref: assertion 'closure->ref_count > 0' failed
```

**Comment by nickeygit** [comment]
> I experimented with a change:

And I've used cgo.Handle (great thing!) - and this allowed to eliminate closures map.
But have not relaxed the issue at all.

The test is still failing and failing very interestingly:
either free() or malloc() are printing random errors.
I've checked glibc source - and all of them - are internal structure of allocated objects lists errors!
Again: it's not about impossibility to allocate or free a normal object. It is about internal allocation mem formatting incorrectness.
This occurs when a process modifies random mem addresses (ie uses uninitialized pointers or so) - this sure can modify glibc allocator internal structure - and generate all the errors we see.
Uninit pointers?
glib/gstreamer - are ok, as native-C MT test works fine.
go-glib/gst - sure, work with pointers, but 1-threaded apps work ok too. If there were such pointers misuses - we'll see them in such apps too.
So, what is it then? Everything else is prod-known and should be pretty stable.

Before, I was looking if any of the 4 libs use Thread Local Storage (TLS) - none do.
Now, I decided to check cgo source itself - and.... OMG:

```
src/runtime/cgo:
libcgo.h:       uintptr *tls;
gcc_linux_amd64.c:void (*x_cgo_inittls)(void **tlsg, void **tlsbase) attribute((common));
gcc_linux_amd64.c:x_cgo_init(G *g, void (*setg)(void*), void **tlsg, void **tlsbase)
gcc_linux_amd64.c:      if (x_cgo_inittls) {
gcc_linux_amd64.c:              x_cgo_inittls(tlsg, tlsbase);
```

CGO itself uses TLS!
TLS - is thread-own mem buffer, which stores objects available only for the owning thread.
Skipping internal TLS work principals - other threads can't access objects allocated by the thread.

And having tons of goroutines migrating between OS threads here and there - **surely** - makes junket from the process memory. And so - not surprising that glibc/malloc()/free() raises their errors.

So, we **must** lock goroutines to threads in multi-go apps.
We must **runtime.LockOSThread()** on every goroutine by definition. This is absolute minimum and required thing we must do in this scenario.

When client thread is only one (or global lock serializes all the cgo calls!) - then cgo runs on one and the same thread always. Guess, they do this intentionally - as they sure know abt TLS in cgo.


I was trying runtime.LockOSThread() at bench() function start few weeks ago - it was greatly relaxing the issue. But not fixing completely (muuuuch less freq. and veeery rarely, but asserts still appear). As it was not fixing the thing completely - I've not proposed it.
Now I see that thread locking is absolutely required.

I've also tried to thread-lock mainloop allocating/Run()ning in a separate goroutine - not helping either. And even running mainLoop.Run() in the main goroutine/thread. Rarely, but asserts are still happening...

**Comment by nickeygit** [comment]
So, cgo is being initialized somewhere earlier at some thread before goroutine is locked to it.
Where?

The only thing I see so far - is here:
```
> ldd ./test1_os_thread
        linux-vdso.so.1 (0x00007fff04f94000)
        libgio-2.0.so.0 => /lib/x86_64-linux-gnu/libgio-2.0.so.0 (0x00007f07a6850000)
        libgobject-2.0.so.0 => /lib/x86_64-linux-gnu/libgobject-2.0.so.0 (0x00007f07a67ed000)
        libglib-2.0.so.0 => /lib/x86_64-linux-gnu/libglib-2.0.so.0 (0x00007f07a66a4000)
        libgstcontroller-1.0.so.0 => /lib/x86_64-linux-gnu/libgstcontroller-1.0.so.0 (0x00007f07a6691000)
        libgstreamer-1.0.so.0 => /lib/x86_64-linux-gnu/libgstreamer-1.0.so.0 (0x00007f07a653e000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f07a632a000)
        libgmodule-2.0.so.0 => /lib/x86_64-linux-gnu/libgmodule-2.0.so.0 (0x00007f07a6323000)
        libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007f07a6307000)
        libmount.so.1 => /lib/x86_64-linux-gnu/libmount.so.1 (0x00007f07a62ba000)
        libselinux.so.1 => /lib/x86_64-linux-gnu/libselinux.so.1 (0x00007f07a628d000)
        libffi.so.8 => /lib/x86_64-linux-gnu/libffi.so.8 (0x00007f07a627f000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f07a6196000)
        libpcre2-8.so.0 => /lib/x86_64-linux-gnu/libpcre2-8.so.0 (0x00007f07a60fc000)
        libunwind.so.8 => /lib/x86_64-linux-gnu/libunwind.so.8 (0x00007f07a60e1000)
        libdw.so.1 => /lib/x86_64-linux-gnu/libdw.so.1 (0x00007f07a602d000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f07a6a31000)
        libblkid.so.1 => /lib/x86_64-linux-gnu/libblkid.so.1 (0x00007f07a5ff2000)
        liblzma.so.5 => /lib/x86_64-linux-gnu/liblzma.so.5 (0x00007f07a5fbe000)
        libelf.so.1 => /lib/x86_64-linux-gnu/libelf.so.1 (0x00007f07a5fa0000)
        libzstd.so.1 => /lib/x86_64-linux-gnu/libzstd.so.1 (0x00007f07a5ee6000)
        libbz2.so.1.0 => /lib/x86_64-linux-gnu/libbz2.so.1.0 (0x00007f07a5ed2000)
```

The ld.so loader itself.
(Ubuntu 22.04)

All the lib names are compiled inside and they are loaded on ELF loading. And guess, cgo is taking some part in the initialization there. And this can happen on some first thread and switch to other threads later. So, cgo first internals are allocated before thread-locking.
These are just my guesses, not checked yet. But thought should describe this.

And the only well-known way to add more flexibility to the init process I know - is to load glib and gstreamer so-libs manually using dlopen()/dlsym()...
This is very different way to load a 3rd-party libraries. Requires much work to remake all the go-gst library internals to load symbol-by-symbol.
So, wondering if you @RSWilli see a better way to get over the cgo init process.

**Comment by RSWilli** [comment]
> And the only well-known way to add more flexibility to the init process I know - is to load glib and gstreamer so-libs manually using dlopen()/dlsym()...

Hard veto, I want to experiment with Full static gstreamer builds statically linked into golang in the future. That means the compiler must be able to include them into the binary statically. No manual dynamic linking.

> So, we must lock goroutines to threads in multi-go apps. We must runtime.LockOSThread() on every goroutine by definition.

Interesting, I never read about this in any cgo documentation, do you have a reference link? The [runtime.LockOSThread docs](https://pkg.go.dev/runtime#LockOSThread) make sense, but go-gst does not depend on these, maybe gstreamer does, but cgo threads should be already locked. (I may be wrong)

> I was trying runtime.LockOSThread() at bench() function start few weeks ago - it was greatly relaxing the issue.

It may be a result of more needed context switches when scheduling the goroutines. I think the HandlerDisconnect crash is a result of some wrong order of unref/ closure invalidation done by go-gst.

**Comment by nickeygit** [comment]
> but cgo threads should be already locked. (I may be wrong)

Nope, this happens on demand only:

CGO normal thread in dump:
```
goroutine 7 [running]:
main.test_func_pipe_and_closure()
        /app/test1_os_thread/test1_os_thread.go:175 +0xb8
main.bench()
        /app/test1_os_thread/test1_os_thread.go:67 +0x1e
created by main.do_tests in goroutine 1
        /app/test1_os_thread/test1_os_thread.go:59 +0x25
```

And at bench() start LockOSThread() added:
```
goroutine 29 [running, locked to thread]:
main.test_func_pipe_and_closure()
        /app/test1_os_thread/test1_os_thread.go:175 +0xb8
main.bench()
        /app/test1_os_thread/test1_os_thread.go:67 +0x25
created by main.do_tests in goroutine 1
        /app/test1_os_thread/test1_os_thread.go:59 +0x25
```

**Comment by nickeygit** [comment]
> I think the HandlerDisconnect crash is a result of some wrong order of unref/ closure invalidation done by go-gst.

We have no Closure Go class/struct wrapper - so no finalizers/Unrefs from Go runtime.
We deal only with *C.GClosure.

**Comment by nickeygit** [comment]
@RSWilli you'll be laughing :)

My test is running for an hour already. No errors/asserts.

I was thinking, I've covered all the goroutines with LockOSThread().
But then a thought come: the finalizers - they are in own goroutines and calls Unref() -> g_object_unref()
So, they touch GST from not-ThreadLocked goroutines.
And I've added locking to all Unref() methods of all the objects both in go-gst/go-glib. Like this:
```
func (v *Object) Unref() {
	runtime.LockOSThread()
	C.g_object_unref(C.gpointer(v.GObject))
}
```

And it's working :)

(very low mem leaking still - for another ticket)

Well, sure it's running a new OS thread per every finalized object - not so cheap thing, but the profit is much greater. Plus, it's possible to change lib design and make unreffing optionally manual for those who care about those extra threads.


You'd ask for a doc why we have to lock cgo to threads - but all I have to show is:
```
/usr/local/go/src/runtime
/usr/local/go/src/runtime/cgo
```
(Source code **is a document**)

Just grep there for "tls' and "thread-local" - and you'll see that it is heavily used there.
Not documented? - well, such specific "feature" looks like a bug then.
Any software can have bugs.

Additionally to runtime.LockOSThread() in bench() and all Unref() methods - had to make instance+handle as key for signals map. Without this not found handlers errors were still appearing.

With these changes - everything is clean and this bug is gone. Multi-threading is working.

Should I prepare PR?

**Comment by nickeygit** [comment]
Oh well, still asserting/crashing sometimes...

Searched the web and found an interesting thing about the problem.

It's a known issue for some graphics libraries (like SDL or OpenGL) - they **require** to be called only from the **main** thread. Not even some same, just main.

Russ Cox presented (at 2012) a proxy template for the problem - described here:
https://golang.google.cn/wiki/LockOSThread

All func calls are being sent to the main goroutine locked to main thread - and executed there.

This is just like the every function global locking I posted before, but even more strict.

Looks like, this is our last resort for multi-threading...

**Comment by RSWilli** [comment]
> So I've added locking to all Unref() methods of all the objects both in go-gst/go-glib. Like this:

The unref is called by the GC mostly and I have a feeling that locking GC to threads has some bigger implications.

> We have no Closure Go class/struct wrapper - so no finalizers/Unrefs from Go runtime.
> We deal only with *C.GClosure.

The unreffing/invalidation of the closures is pretty complicated and a bit obfuscated by the fact that CGo calls the unrefs on our side. On closure invalidation (either by [disconnecting the handler](https://github.com/go-gst/go-glib/blob/main/glib/gobject.go#L323) or other ways) cgo calls the [attached finalizer](https://github.com/go-gst/go-glib/blob/main/glib/connect.go#L32) (defined [in the C part](https://github.com/go-gst/go-glib/blob/main/glib/glib.go.h#L153-L157)) that [removes the closure from our map](https://github.com/go-gst/go-glib/blob/main/glib/connect.go#L114).

I have a feeling that this invalidation gets called multiple times for the same closure, which throws the asserts. In a racy case it may happen that the [returned closure is a nil pointer](https://github.com/go-gst/go-glib/blob/main/glib/gobject.go#L322) (because we do not check if the map access was successful) which then crashes. I cannot prove this yet though.

**Comment by nickeygit** [comment]
> The unref is called by the GC mostly and I have a feeling that locking GC to threads has some bigger implications.

Yes, but anyway, this is not completely helping.


> ... cgo calls the attached finalizer that removes the closure from our map.

Right. Just another finalizer form.
Have played around this all much. Even removing the finalizer attaching (C._g_closure_add_finalize_notifier) is not fixing the issue.



> I have a feeling that this invalidation gets called multiple times for the same closure, which throws the asserts. In a racy case it may happen that the returned closure is a nil pointer.

Yes, we already discussed that code place. And I have a go-glib version with the maps replaced with cgo.Handle - this all doesn't help.
The issue is some deeper.

Please consider this.
The test can crash in **many** different ways. Sometimes it prints malloc-related errors too. Like this:
```
malloc_consolidate(): invalid chunk size
malloc(): unaligned fastbin chunk detected
malloc(): smallbin double linked list corrupted
free(): invalid next size (fast)
malloc(): unaligned tcache chunk detected
```
I've gathered them in a few minutes running the test again and again. It's easy to repeat this and I frequently see these errs - so it's not something rare about the issue.

Do we see the errors in malloc/free manual pages? - No. They are not "normal errors". It's impossible to get them just calling malloc/free usually.

I got glibc-2.39 source and checked for them - they are in places where internal glibc mem allocation structures are checked (ie some sort of file system structure in memory).

malloc/malloc.c:
```
      if (__glibc_unlikely (pp != NULL && misaligned_chunk (pp)))       \
    malloc_printerr ("malloc(): unaligned fastbin chunk detected"); \
```
```
      if (__glibc_unlikely (bck->fd != victim))
        malloc_printerr ("malloc(): smallbin double linked list corrupted");
```
and so on.

It's literally impossible to damage glibc internals so hard just by malloc()'ing wrong buffer size (the only one parameter to the function) or free()ing wrong address or free() twice. 
So, it's not about our pointers usage.

From other hand - such horrible errors can be caused by using on a thread Thread-local-storage of another thread or so (like put memory map of one process to another). Which is possible without locking goroutines to threads or so. And this explains why locking - makes the issue muuuch relaxed and crashing much less frequently (not 100% though).

So, the issue is much deeper. Something in cgo/multithreading model.
And fixing it - is just other level of time spending/hardness.

So, we should accept the cgo complexity and implement proxy-pattern for calls from non-main threads.
Like Russ Cox advised for near same issue with other TLS-using libs.
(I don't like to refer authoritative ppl - but I've already debugged enough to see the complexity of the issue and that the advise is a way to fix/avoid it, having cgo as is)

We can have 2 switchable modes:
* direct C/glib/gst calls (like it is now): the caller just touches C.g/gst_  functions
* proxy: the caller sends requests to main thread and it does the C.g/gst_ calls for him

Existing library users will not see any changes.
While multithreading apps will be able to work.

Please say what you think.

**Comment by RSWilli** [comment]
@nickeygit sorry if it sounds like I want to block your suggestion. It is just that you are suggesting a big change and I cannot fully grasp the consequences.

I dug around in the gotk3 bindings (from which `go-glib` was orignally forked) and found the following issue that suggests that your suspicion may be correct: https://github.com/gotk3/gotk3/issues/863

The issue suggests that sending the calls to the main thread can be done via `IdleAdd` in there bindings. Not sure if we have a similar thing in our bindings.

The other question I have is which calls would we need to protect? Surely it isn't every `C.gst_` call, because I have some production applications that are running for months just fine, but with only one running pipeline. It may be that this completely depends on the pipeline.

**Comment by nickeygit** [comment]
> @nickeygit sorry if it sounds like I want to block your suggestion. It is just that you are suggesting a big change and I cannot fully grasp the consequences.

Np, there are diff ways to do that, including [near-]zero changes to existing code. (see below)



> I dug around in the gotk3 bindings (from which `go-glib` was orignally forked) and found the following issue that suggests that your suspicion may be correct: [gotk3/gotk3#863](https://github.com/gotk3/gotk3/issues/863)

Oh God! Yes, it's same thing.


> The issue suggests that sending the calls to the main thread can be done via `IdleAdd` in their bindings. Not sure if we have a similar thing in our bindings.

We not. Sure, we could add it to our go-glib - it's C.g_idle_add_full() under it. But.
This means all the secondary threads still will touch cgo bridge with all its uncertain probs - which we are trying to avoid now.
So, we should minimize places where go <-> cgo bridge is crossed.
Only the main thread must be able to do this.
But how to accept proxy requests then? As main thread should run blocking Mainloop.Run() -> C.g_main_loop_run().
There is another way to process events in glib:
g_main_context_iteration():
```
Runs a single iteration for the given main loop. ...
```
Like _run(), but without the cycle. Exactly what we need.
So, the main goroutine/thread (guess, locked to thread) will be able both to receive requests from a task channel, process the requests and handle glib main loop events.


> The other question I have is which calls would we need to protect? Surely it isn't every `C.gst_` call, because I have some production applications that are running for months just fine, but with only one running pipeline. It may be that this completely depends on the pipeline.

Guess - as mentioned above, we have to protect cgo bridge itself. Only main thread should be able to call C.xxxx. So, go-gst/go-glib functions that don't call cgo - don't need the protection. It's hard to find such ones though (as go-gst is a wrapper itself :) - so we'd better think to protect them all. (below - how to make this process simpler)

And now, how we can do that without big existing code changes.
Following SOLID principle SRP (Single Responsibility Principle) - one function or class or module - should have only 1 reason to change (ie 1 responsibility).
Current go-gst implementation (package go-gst/gst/) does have it: to be a bridge to gstreamer at all.
Multi-threading support using a proxy passing requests between threads - is another thing/responsibility. So, it should be placed in another package, like go-gst/mt/ or go-gst/multithread/ or so.
And this go-gst/mt/ package will just use go-gst/gst/

go-gst/mt/gst/ -> uses -> go-gst/gst/
go-gst/mt/glib/ -> uses -> go-glib/glib/

So, MT apps will have the same API as right now. Just need to change:
import ...... go-gst/gst/
to
import ...... go-gst/mt/gst/

and as well for go-glib.
Or even add the /mt/ package to go-glib - but this blocks potential switching to original gotk3 :)
(but we still need to fix go-glib for that instance+handle keys for closure map)

While migrating to gotk3 will be easy enough:
go-gst/mt/glib/ -> uses -> gotk3/glib/

The proxy implementation is pretty simple: channel, requests, readiness response channel etc.

And sure, it's huge number of functions/methods to wrap. So, it would be faster to write a converter from current go-gst/gst/ API - to proxy-wrapped add it to the project. (then run/update it once a year or less - as Glib/Gstreamer updates).

**Comment by RSWilli** [comment]
> Only main thread should be able to call C.xxxx.

what is the definition of "main thread" here? Is it the thread with the glib main loop or the main application thread? These might not be the same. Since go handles goroutines and multiplexes them onto threads some of the definition might be very go specific.

> And sure, it's huge number of functions/methods to wrap.

I'd like to avoid that because it adds a lot of complexity to this already pretty big library, which could scare away contributors. I think we should figure out a general way to do it in the bindings. The [GNOME docs](https://developer.gnome.org/documentation/tutorials/threading.html) state that threading should be avoided completely when interacting with Glib. Maybe the aforementioned `g_idle_add` is the way to go here.

If we have to touch the current code a lot then be it. I just want to be sure that this is actually the problem we are facing and the best fix for it.

**Comment by nickeygit** [comment]
> what is the definition of "main thread" here? Is it the thread with the glib main loop or the main application thread? These might not be the same. Since go handles goroutines and multiplexes them onto threads some of the definition might be very go specific.

In a go app, main thread - is the first process thread running the main.main goroutine.
But here, I mean the thread, which calls cgo/gst code. This can be another thread too I think.
At least, the very first idea to add global locking to every function, had GST mainloop running in other goroutine - potentially on other thread. And that arch was working, were no asserts/crashes. So, I think (think) - all the cgo code was using just some same thread always (as there was only 1 cgo request at a moment).
So, we can talk not strictly about the main thread, but about a thread which calls cgo.
So the term could be "cgo thread" or so.


> I'd like to avoid that because it adds a lot of complexity to this already pretty big library, which could scare away contributors. I think we should figure out a general way to do it in the bindings.

Well, ok. Then every func/method will have to be modified.
Existing methods should be untouched and just renamed to non-published, like
SetProperty -> setProperty

and SetProperty() itself will contain the routing code - how to handle the call:
here or send to cgo-thread proxy.
Something as small as possible will be added, like:
```
func (v *Object) SetProperty(......) {
  mt_lock. Rlock()
  defer mt_lock.Runlock()

  if !mt_enabled {
    return setProperty(....)
  }
  return setProperty_mt(....)
}
```
(So, global mt_enabled flag will allow single-threaded apps to run as before).

And setProperty_mt() must be created (pretty templated one) - to create proxy request, send and wait for response.

Alternatively, we could use a single universal method to send such MT requests,
but will have to use "unsafe" package to dig in runtime internals to find out called:
method name, receiver params and values to return.

Well, both approaches are complex in diff ways.


> The [GNOME docs](https://developer.gnome.org/documentation/tutorials/threading.html) state that threading should be avoided completely when interacting with Glib.

Just brilliant advise to avoid threading, really :) But, unfortunately, not for multimedia multiusers network servers... It's not a one-user desktop app...


> Maybe the aforementioned `g_idle_add` is the way to go here.

This is a way to avoid cgo bridge touching/issues using...  cgo bridge itself :)
Many threads calling a cgo function over cgo bridge...  I guess, I was facing this for few weeks already :)
And that happy guy from gotk3 issue - could just not test enough this approach - for up to 3 goroutines - current test is pretty rare to crash...

So, guess, we should do cross-thread communication inside Go only. And only Cgo-thread then touch cgo-bridge.


> If we have to touch the current code a lot then be it. I just want to be sure that this is actually the problem we are facing and the best fix for it.

Absolutely. I'd like to fix this issue with one runtime.LockOSThread() or so too :) but things appear more complex...
The best fix would be - for sure - in Cgo itself, so it would allow to make cross-thread calls. But well, it is much more complex to find, fix and discuss in Go-chat... So, we should just avoid this crashy cgo "feature" for now.
And avoiding is simple:
the Go code deal inside all its goroutines what to ask from cgo/gst - and dedicated cgo-thread does the call.

**Comment by RSWilli** [comment]
@nickeygit I talked to @sdroege and he agrees that adding a global lock (essentially making `go-gst` single threaded) is a bad idea.

I created https://github.com/RSWilli/go-gst-issue-94 as a reproduction repo.

**Comment by nickeygit** [comment]
> @nickeygit I talked to @sdroege and he agrees that adding a global lock (essentially making `go-gst` single threaded) is a bad idea.

We are not about adding a global lock, but global task channel (which implies locking though) with single calling thread.

Well, I see @sdroege is a rust-man - no Go repositories in his account - while we face a Go/Cgo issue here.
And making go-gst optionally/switchably (!!!) single-threaded - is a workaround for the broken cgo.

Sure, he doesn't know about Cgo - so reasonably responded that it's bad idea.
It would be bad idea with multi-threaded Cgo. But we have only single-threaded Cgo in fact.

I already have proposed 2 fixes/workarounds for our case:
* global lock in every call (optionally switchable to have old-known behavior) - this looks incomplete for me now - as doesn't lock the caller to a single thread.
* proxy-pattern (again, optionally switchable to have old-known behavior) to have single locked caller thread. This is the only way to work with broken cgo.

There were no other ideas about how to get this test working though...

Fix cgo first?

**Comment by RSWilli** [comment]
> Well, I see sdroege is a rust-man

maybe take a look at his alt account at https://gitlab.freedesktop.org/slomo :) He is one of the main gstreamer developers and offered to help.

I've been told that gstreamer itself is designed to be very thread safe. He is unsure if TLS has anything to do with this issue. I'd like to investigate further before rushing to a conclusion that affects every function call in this repository.

**Comment by nickeygit** [comment]
> maybe take a look at his alt account at https://gitlab.freedesktop.org/slomo :) He is one of the main gstreamer developers and offered to help.

Yep, he is good :) And his help is welcome for sure.



> I've been told that gstreamer itself is designed to be very thread safe.

It really is thread-safe. That's true (I was looking at its code for these weeks too - mutexes are everywhere).
But it doesn't matter when executing machine drops/skips values/data/etc any moment it wants :)

So sure, please feel free to investigate.

**Comment by sdroege** [comment]
@nickeygit I just want to make sure this is not actually a bug elsewhere.

Also it seems strange that cgo would be completely not thread-safe at all: that would make Go rather useless for the last decade or more since multi-processor systems became common :upside_down_face: 

Do you have some details about the exact limitations? cgo itself is thread-safe as long as you don't access the same value from multiple threads? But you e.g. can't call the same closure from multiple threads at once?

What exactly is happening in this specific case that breaks cgo assumptions? You wrote something about TLS, that would be TLS on the Go runtime side?

**Comment by RSWilli** [comment]
@sdroege my suspicion is rather that the invalidation of the signal handlers and their closures on the go side is wrong. I suspect that some already invalidated closures get invalidated again in rare cases.

**Comment by nickeygit** [comment]
> @nickeygit I just want to make sure this is not actually a bug elsewhere.
> 
> Also it seems strange that cgo would be completely not thread-safe at all: that would make Go rather useless for the last decade or more since multi-processor systems became common 🙃

I'm not saying "completely". Just some circumstances that trigger some bug.
This is just suspicion on cgo - as I could not find the reason in the go-gst/go-glib code.
(Maybe somebody else can though)

> Do you have some details about the exact limitations? cgo itself is thread-safe as long as you don't access the same value from multiple threads?

Officially yes, it's MT safe. Limitations? - Guess, something related to Thread Local storage - as CGO uses the technology. It has a bug here or some other glib/gstreamer/wrappers use it too - causing some mess in thread-related pointers (so threads touch not their memories).

> But you e.g. can't call the same closure from multiple threads at once?

This is not happening in this bug testing code.
And it requires to be more specific here: OS threads or goroutines.
Every goroutine in the test - touch only own closures. As well as shared closures map - but it's properly mutex-protected. Handler IDs has incomplete keys though - but this is another bug and has simple fix (unfortunately, with this bug fixed - the #94 crashing is still happening).

> What exactly is happening in this specific case that breaks cgo assumptions? You wrote something about TLS, that would be TLS on the Go runtime side?

Guess, I've already described all I new at the moment about this.

* I see no go-gst/go-glib code reasons to mess memory so hard - that **glibc** (not glib!) memory allocator's internal structures get broken ( https://github.com/go-gst/go-gst/issues/94#issuecomment-2252524988 ). Plus, again, gotk3/gotk3 has the same multi-goroutines issue.

* And I see a huge potential in TLS thread-related pointers to make such a mess. Just imagine: one thread - using own TLS pointer/buffer has allocated an object there (with normal malloc()) - and with relative pointer (address shift from TLS buffer start to the object) this object can be accessed safely from this thread only (using normal absolute pointer - it's accessible for all other threads though). And some code (goroutine), having the shift, wants to access the object. Once it run from the wrong OS thread (and so - with wrong TLS) - the shift would point to a wrong address and modifying it can cause mess there. And it can be internal glibc mem allocator structures. With the damaged structures - when glibc checks them again during free() (the most of crash stacks finish with free()!) - it finds **many different** corruptions in them - and prints warnings about to stdout.
And Go's scheduler and mapping G-Goroutines with P-processors (software) to M-os-threads - make it possible to run a goroutine on a different threads.

Making every goroutine to be locked to an OS thread (runtime. LockOSThread() call) - **greatly** works around the issue (it requires much more running time to see errors). And it's reasonable: if everybody is locked to its own Thread/TLS - then such corruptions don't happen in such amounts.
And the code is still Multi-threaded - so runs in parallel. With all the go-gst locks/maps. So, the bug unlikely can be explained just by signals/closures invalidation incorrectness.

I recommend to look into Go source code directory (any fresh Go version; I've repeated this crash on 1.20, 1.21, 1.22):
/usr/local/go/src/runtime 

and search for "TLS" (case in-sensitive).
You'll see massive usage of it. So, Go runtime and cgo (subdir there) - greatly depends on TLS.
Next, you should note the code complexity and uncleanness.
Well, have to say that: Golang source code is not clean - in terms of "Clean Code" author Robert Martin - aka Uncle Bob.
Many describing comments (while it's code works, not comments).
Huge functions doing **many** things - while a function should do 1 thing only, and do it well.
Crossing with assembler code from different platforms etc.
Sure, there are many tests for all this. But the code was not written using TDD obviously (it would be much cleaner otherwise) - so unlikely all the run-cases are tests-covered. So, such hard-crashing consequences are not so unreasonable.

But sure I can be wrong and the bug is somewhere in go-gst-glib.
Have another way to try to track it down:
start to remove go-gst/go-glib libs code keeping the issue happening.
Earlier or later - there should left few code clearly pointing to the reason.
Just a thought.

**Comment by sdroege** [comment]
> I see no go-gst/go-glib code reasons to mess memory so hard - that glibc (not glib!) memory allocator's internal structures get broken

That looks like any kind of memory corruption unfortunately. The problem could be anything really.

> And I see a huge potential in TLS thread-related pointers to make such a mess. [...]

I don't see how that can cause problems here. Any call into a C function from Go also returns from that function in the very same thread and without another goroutine running on the same thread before returning, or else nothing would work and it would be unsafe to call the majority of C functions (you'd basically be limited to async-signal-safe functions, which are not... many).

> Making every goroutine to be locked to an OS thread (runtime. LockOSThread() call) - greatly works around the issue (it requires much more running time to see errors). And it's reasonable: if everybody is locked to its own Thread/TLS - then such corruptions don't happen in such amounts.

That indeed hints at a thread-safety problem somewhere but it feels at this point like this is most likely in go-gst somewhere.

As a side note, serializing all GStreamer FFI code through a single thread basically brings you into the situation the Python GStreamer bindings have thanks to the GIL. It's a serious performance problem for any GStreamer application that doesn't just build pipelines but also manipulates data flow (pad probes, custom element implementations, etc) or otherwise gets called from streaming threads.

**Comment by nickeygit** [comment]
> > And I see a huge potential in TLS thread-related pointers to make such a mess. [...]
> 
> I don't see how that can cause problems here. Any call into a C function from Go also returns from that function in the very same thread and without another goroutine running on the same thread before returning, or else nothing would work and it would be unsafe to call the majority of C functions (you'd basically be limited to async-signal-safe functions, which are not... many).

Sounds good and reasonable. And the cgo runtime really looks like does this.
Just an advice to look at CGO entrypoint function (the one appearing in the crash stacks mostly):
runtime.cgocall()
in runtime/cgocall.go file.
It really intents to lock itself to the running thread during the call - so other G-s not be scheduled here.
But again, the code it not clean - and so, intentions are not always mean actual actions...


> > Making every goroutine to be locked to an OS thread (runtime. LockOSThread() call) - greatly works around the issue (it requires much more running time to see errors). And it's reasonable: if everybody is locked to its own Thread/TLS - then such corruptions don't happen in such amounts.
> 
> That indeed hints at a thread-safety problem somewhere but it feels at this point like this is most likely in go-gst somewhere.

Go "userlevel" code don't work with OS threads durectly, only Gouroutines. So, if a G is locked to a thread - has nothing to do for the Go application code. It is async anyway and all the shared structs must be properly locked. But with runtime.LockOSThread() they are **more properly** locked somewhy... ;)
And again, please feel free to read the function comments - to have more info to think:
```
...
// A goroutine should call LockOSThread before calling OS services or
// non-Go library functions that depend on per-thread state.
```
And TLS **is** a per-thread state.
Does "userlevel" Go code has anything to do with threads TLS?


> As a side note, serializing all GStreamer FFI code through a single thread basically brings you into the situation the Python GStreamer bindings have thanks to the GIL. It's a serious performance problem for any GStreamer application that doesn't just build pipelines but also manipulates data flow (pad probes, custom element implementations, etc) or otherwise gets called from streaming threads.

Yeah, I know single-threading is just a workaround. The fix still should be found.
But my thought that the fix is not related to this go-gst project.

**Comment by RSWilli** [comment]
> also returns from that function in the very same thread and without another goroutine running on the same thread before returning

The fun part is that this is entirely possible with goroutines itself, but cgo seems to pin the calling goroutine (aka `G`) to the thread (aka `M`).

References:

* https://go.dev/src/runtime/HACKING terminologies in the runtime: Gs, Ms, Ps
* https://github.com/golang/go/blob/master/src/runtime/cgocall.go#L26-L28 Cgo calling thread is unlocked after the call, meaning it must have been locked before?
*  https://github.com/golang/go/blob/master/src/runtime/cgocall.go#L329 when C calls a go function we stay locked on this thread until the calling goroutine returns control back to c (this is the case for signal handler callbacks)

**Comment by sdroege** [comment]
Yeah well that's the whole point of (g|c)oroutines, but if you extend that to unsuspecting C code there's very little code you can actually safely call. Thanks for pointing out those parts of the cgo code :)

**Comment by RSWilli** [comment]
There is still the question whether we are currently racing against gstreamer while invalidating a signal handler and their closure.

@sdroege if you take a look at [`HandlerDisconnect`](https://github.com/go-gst/go-glib/blob/main/glib/gobject.go#L317-L330), is there anything that should be done differently? I think we currently have two problems in this code:

* the signal handler map is wrong, because handler IDs are unique per object and not globally unique
* the closure sometimes is already invalidated when we call `g_closure_invalidate`

**Comment by RSWilli** [comment]
@sdroege how do the ownerships and finalizers of the closure interact with the signals? We are using [g_signal_connect_closure](https://docs.gtk.org/gobject/func.signal_connect_closure.html) which states:

> If closure is a floating reference [...], this function takes ownership of closure.

Does that mean that if we don't call [`g_closure_sink`](https://docs.gtk.org/gobject/method.Closure.sink.html) then Glib takes care of the invalidation of the closure, when the signal is detached? Or do we need to invalidate the closure explicitly after detaching the signal?

**Comment by sdroege** [comment]
Invalidation is unrelated to ownership. GLib takes care of invalidation in any case. What this means is that you don't need to call `g_closure_unref()` on the closure if it's still floating because the function takes ownership of it in that case (if it's not floating anymore then you need to unref it yourself!).

Similar to other things with floating references, you should really make sure to sink them on the bindings side immediately to avoid all kinds of problems.

**Comment by RSWilli** [comment]
Okay so if I understand you correctly, then our current situation is the following:

1. we create a floating closure
2. we pass the closure to the signal handler, which then takes ownership
3. when detaching the signal, the steps are:
    - we detach the signal, making glib invalidate the closure (?)
    - we invalidate the closure

You are suggesting that we should sink the closure, which would mean that we need to keep that reference around to be able to unref it after detaching (and invalidating ?).

Did I miss something?

**Comment by sdroege** [comment]
The current situation is wrong then. After disconnecting the `GClosure`, it's gone and you must not invalidate it anymore (and GLib already did that for you anyway). That's a classic use-after-free bug.

> You are suggesting that we should sink the closure [...]

That's the general rule for GObject bindings and any kind of floating references (including GstObjects, GVariant, etc). If you do it that way, you still have the `GClosure` after disconnecting, can call API on it and then unref it when you don't need it anymore.

**Comment by RSWilli** [comment]
>  If you do it that way, you still have the GClosure after disconnecting

(for signals at least) I like the approach better where the closure is automatically invalidated and unreffed by glib, because we create one GClosure per attached signal. That way we do not need to track the closures and the signal handlers. If we want to support that the user attaches GClosures themselves, then they would need to take care of the ownership themselves.

I'd like to experiment a bit with storing the callback function itself directly in the closure, making our currently globally mutex protected maps completely obsolete. I will have to check against cgo's calling conventions regarding go pointers, but in theory this should be possible.

**Comment by sdroege** [comment]
Storing it inside the `GClosure` sounds like a good idea, yes

**Comment by RSWilli** [comment]
I had some time to play around with this. I did the following:

```c
extern void goMarshal(GClosure *, GValue *, guint, GValue *, gpointer,
                      GValue *);

static GClosure *_g_closure_new(gpointer data)
{
  GClosure *closure;

  closure = g_closure_new_simple(sizeof(GClosure), data);
  g_closure_set_marshal(closure, (GClosureMarshal)(goMarshal));
  return (closure);
}

extern void removeClosure(gpointer, GClosure *);

static void _g_closure_add_finalize_notifier(GClosure *closure, gpointer data) {
  g_closure_add_finalize_notifier(closure, data, removeClosure);
}
```

This way the closure keeps track of the callbacks via the data attribute. This cleans up a lot of our code, which I really like. On the other hand I still got it to crash sometimes with the test pipeline (although not yet while running a debugger).

The problem is that sometimes the `goMarshal` receives a closure with their data unset. I have not figured out where this originates yet.

I also had to pass data to `g_closure_add_finalize_notifier` because the finalizer callback had the same issue, where the data in the closure was gone. Maybe this is related, but I couldn't find why. @sdroege can you help?
